cmd/mysterium_client/command_run/command.go

```diff
@@ -5,90 +5,94 @@ import (
+	nats_discovery.Bootstrap()
+	openvpn.Bootstrap()
 
-	session, err := cmd.MysteriumClient.SessionCreate(options.NodeKey)
-	if err != nil {
-		return err
-	}
-	proposal := session.ServiceProposal
-
-	dialogEstablisher := cmd.DialogEstablisherFactory(consumerId)
-	cmd.dialog, err = dialogEstablisher.CreateDialog(proposal.ProviderContacts[0])
-	if err != nil {
-		return err
-	}
+	mysteriumClient := server.NewClient()
 
-	vpnSession, err := vpn_session.RequestSessionCreate(cmd.dialog, proposal.Id)
-	if err != nil {
-		return err
-	}
+	keystoreInstance := keystore.NewKeyStore(options.DirectoryKeystore, keystore.StandardScryptN, keystore.StandardScryptP)
 
-	vpnConfig, err := openvpn.NewClientConfigFromString(
-		vpnSession.Config,
-		options.DirectoryRuntime+"/client.ovpn",
-	)
-	if err != nil {
-		return err
-	}
+	identityManager := identity.NewIdentityManager(keystoreInstance)
 
-	vpnMiddlewares := append(
-		cmd.vpnMiddlewares,
-		bytescount_client.NewMiddleware(cmd.MysteriumClient, session.Id, 1*time.Minute),
-	)
-	cmd.vpnClient = openvpn.NewClient(
-		vpnConfig,
-		options.DirectoryRuntime,
-		vpnMiddlewares...,
-	)
-	if err := cmd.vpnClient.Start(); err != nil {
-		return err
+	dialogEstablisherFactory := func(identity identity.Identity) communication.DialogEstablisher {
+		return nats_dialog.NewDialogEstablisher(identity)
 	}
 
-	router := tequilapi.NewApiRouter()
-
-	keystoreInstance := keystore.NewKeyStore(options.DirectoryKeystore, keystore.StandardScryptN, keystore.StandardScryptP)
-	idm := identity.NewIdentityManager(keystoreInstance)
-	endpoints.RegisterIdentitiesEndpoint(router, idm)
+	vpnClientFactory := configureVpnClientFactory(mysteriumClient, options.DirectoryRuntime)
 
-	endpoints.RegisterConnectionEndpoint(router, client_connection.NewManager())
+	vpnManager := client_connection.NewManager(mysteriumClient, dialogEstablisherFactory, vpnClientFactory)
 
-	cmd.httpApiServer, err = tequilapi.StartNewServer(options.TequilaApiAddress, options.TequilaApiPort, router)
+	httpApiServer, err := tequilapi.NewServer(options.TequilaApiAddress, options.TequilaApiPort)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	return nil
+	return &CommandRun{
+		mysteriumClient,
+		identityManager,
+		vpnManager,
+		httpApiServer,
+	}, nil
+}
+
+func (cmd *CommandRun) Run() {
+	router := tequilapi.NewApiRouter()
+	endpoints.RegisterIdentitiesEndpoint(router, cmd.identityManager)
+	endpoints.RegisterConnectionEndpoint(router, cmd.connectionManager)
+	cmd.httpApiServer.StartServing(router)
 }
 
 func (cmd *CommandRun) Wait() error {
 	return cmd.httpApiServer.Wait()
 }
 
 func (cmd *CommandRun) Kill() {
-	cmd.dialog.Close()
-	cmd.vpnClient.Stop()
 	cmd.httpApiServer.Stop()
 }
+
+func configureVpnClientFactory(mysteriumApiClient server.Client, vpnClientRuntimeDirectory string) client_connection.VpnClientFactory {
+	return func(vpnSession *session.VpnSession, session *dto.Session) (openvpn.Client, error) {
```

Comment:
- Very strange to inject all `*session.VpnSession` but need just `sessionConfig`
- Very strange to inject all `*dto.Session`, then You need just `sessionId`

Extra tradeoff:
- Then you try to pass by reference, which leads to unsafe `nil` passing

Reply:
It's a result of breaking a much bigger function and trying to isolate creating vpn client from dialog passed configuration. But a few moments - dto.Session should disappear at all, as vpnSession should contain all config and ids needed to construct vpn client. Secondly our only supported transport carrier is VPN, but this factory would be a good place to create a specific transport connection negotiated between client and node, i.e. WebRTC connection or anything else. VpnSession is bad naming as it's esentially an agreement between client and node - what kind of service will be used, it could be not only VPN.
As there is single place that calls this method - caller guarantees that nils will not be passed. Again - why to make extra copy of those two structures?